src/paths/floyd_warshall.c

```diff
@@ -140,3 +142,132 @@ igraph_error_t igraph_distances_floyd_warshall(
 
     return IGRAPH_SUCCESS;
 }
+
+igraph_error_t igraph_distances_floyd_warshall_tree_speedup(
+        const igraph_t *graph, igraph_matrix_t *res,
+        const igraph_vector_t *weights, igraph_neimode_t mode) {
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_bool_t in = false, out = false;
+
+    if (weights && igraph_vector_size(weights) != no_of_edges) {
+        IGRAPH_ERROR("Invalid weight vector length.", IGRAPH_EINVAL);
+    }
+
+    if (! igraph_is_directed(graph)) {
+        mode = IGRAPH_ALL;
+    }
+
+    switch (mode) {
+    case IGRAPH_ALL:
+        in = out = true;
+        break;
+    case IGRAPH_OUT:
+        out = true;
+        break;
+    case IGRAPH_IN:
+        in = true;
+        break;
+    default:
+        IGRAPH_ERROR("Invalid mode.", IGRAPH_EINVAL);
+    }
+
+    if (weights && igraph_vector_is_any_nan(weights)) {
+        IGRAPH_ERROR("Weight vector must not contain NaN values.", IGRAPH_EINVAL);
+    }
+
+    IGRAPH_CHECK(igraph_matrix_resize(res, no_of_nodes, no_of_nodes));
+    igraph_matrix_fill(res, IGRAPH_INFINITY);
+
+    for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+        MATRIX(*res, v, v) = 0;
+    }
+
+    for (igraph_integer_t e=0; e < no_of_edges; e++) {
+        igraph_integer_t from = IGRAPH_FROM(graph, e);
+        igraph_integer_t to = IGRAPH_TO(graph, e);
+        igraph_real_t w = weights ? VECTOR(*weights)[e] : 1;
+
+        if (w < 0) {
+            if (mode == IGRAPH_ALL) {
+                IGRAPH_ERRORF("Negative edge weight (%g) found in undirected graph "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            } else if (to == from) {
+                IGRAPH_ERRORF("Self-loop with negative weight (%g) found "
+                              "while calculating distances with Floyd-Warshall.",
+                              IGRAPH_ENEGLOOP, w);
+            }
+        }
+
+        if (out && MATRIX(*res, from, to) > w) MATRIX(*res, from, to) = w;
+        if (in  && MATRIX(*res, to, from) > w) MATRIX(*res, to, from) = w;
+    }
+
+    igraph_matrix_int_t successors;
+    igraph_stack_int_t stack;
+
+    IGRAPH_STACK_INT_INIT_FINALLY(&stack, no_of_nodes);
+
+    IGRAPH_CHECK(igraph_matrix_int_init(&successors, no_of_nodes, no_of_nodes));
+    IGRAPH_FINALLY(igraph_matrix_int_destroy, &successors);
+
+    for (igraph_integer_t v=0; v < no_of_nodes; v++) {
+        for (igraph_integer_t u=0; u < no_of_nodes; u++) {
+            /* the second vertex on the shortest path from u to v */
+            MATRIX(successors, u, v) = v;
```

Comment:
`succcessors[i,j]` corresponds to $\pi_{ij}$ in the paper in Algorithm 2, correct? Line 1 reads $\pi_{ij} := i$. However, here you are setting $\pi_{ij} = j$. Is there a problem here?

Reply:
It's a bit confusing to follow because the notation here is different than in the paper. I can see that this is in part because you started with the existing plain Floyd–Warshall code in igraph, and kept the iterator names from there. But it would be much easier to follow whether the implementation matches the paper if we used the same notation as the paper.

It seems to me that with the assumption that `successors[j,i]` = $\pi_{ij}$ (i.e. transpose of what I wrote above), the line below saying `MATRIX(successors, i, j) = MATRIX(successors, k, j);` still does not match the paper.

Could you please refactor the notation to match the paper, so it's easier to check the algorithm?

Reply:
> Could you please refactor the notation to match the paper, so it's easier to check the algorithm?

I can refactor it to match the paper if you think the following wouldn't cause some considerable performance drawbacks.
The reason I didn't follow the original algorithm is this comment:
 //  /* Iteration order matters for performance!
        //  First j, then i, because matrices are stored as column-major. */

Exact reasons why the order matters is not clear to me, though.